### ISSUE TITLE
Tabs versus spaces mess up vertical scrolling

### ISSUE THREAD

**Issue by mkrenik** [snaptoken/kilo-tutorial#16]
I am using this project as a way to get more familiar with Vim. I found that when I implemented vertical scrolling, artifacts from neighboring lines were getting copied into the white spaces of other lines. I'm still looking into how exactly this happened. But I was about to fix the issue by changing my .vimrc file to redefine the tab key to insert spaces instead of inserting an actual tab symbol that appeared as multiple spaces.

**Comment by mkrenik** [comment]
I'm pretty confident that I have identified the issue. The function `editorRefreshScreen()` doesn't clear the terminal and rewrite from scratch. It assumes that everything in the new buffer will override everything in that is already printed (remember we use `\x1b[K` to clear everything to the right of the cursor). The issue is that when a tab character is written to the terminal, it doesn't overwrite, but it advances the cursor 4 (or another amount) spaces. The characters that were already written in those 4 spaces aren't overwritten by spaces, they persist. 

For example, lets say you have a terminal that is 25 rows. You have "Jimmy, how are you" written on line 25 and "\tI am fine" written on line 26. When you scroll down, Line 24 now reads "Jimmy, how are you, and Line 25 now reads "JimmI am fine". 

Here are some possible fixes. Keep in mind that I haven't gotten through the entire project, so I don't know which fix is best given how the system is implemented later on. 

- Edit the editorOpen function so that whenever it comes across a tab, it replaces it with multiple spaces. 
- Clear the terminal before refreshing the screen (this might lead to some flickering problems)
- Change some terminal setting such that when a tab is written to the STDOUT, it uses white spaces instead of just advancing the cursor. This is probably the optimal solution, but I'm not familiar enough to know if this setting actually exists.

**Comment by yjerem** [comment]
Tab rendering is fixed starting on step 80. It uses a more complicated fix, as you'll see when you get there. But maybe I should put a note in the vertical scrolling section explaining how to fix it temporarily, using one of your suggestions...

Sorry, I didn't realize how bad the editor would look for people who use a lot of tabs, between steps 69-80. Thanks for bringing my attention to the problem.

**Comment by mkrenik** [comment]
I figured that with how thorough you were, that there might have been a fix later, but I figured it'd be instructive to look into it myself. 

If not a temporary solution, I'd recommend putting in a note that says "you might see this issue..." just so people don't wonder what is going on. I spent some time looking through to see if I had made a mistake earlier in the tutorial. 

In any case, thanks for this guide. I'm really enjoying it and looking forward to doing more!
